### DISCUSSION TITLE
How to open a file in mojo?

### DISCUSSION THREAD

**Discussion by amarouane-ABDELHAK** [modular/modular#732] [General]
I am trying to read a file using mojo but didn't find any example in the getting started documentation

**Comment by rarebreed** [comment]
I don't believe there is a way to do this through any mojo library.  but you can do it through python

```
Welcome to Mojo! 🔥
Expressions are delimited by a blank line.
Type `:mojo help` for further assistance.
1> let bn = Python.import_module("builtins")
2. let f_hand = bn.open("learning.mojo", "r")
3. let body = f_hand.read()
4. print(body)
5. 
from python import Python

fn run(cmd: String) raises -> None:
    let sp = Python.import_module("subprocess")
    let result = sp.run(
        ["ls", "-al"],
        None,     # stdin
        None,     # input
        sp.PIPE,  # stdout,
        sp.PIPE,  # stderr
        True,     # capture_output
    )
    print(result.stdout)
```

Since I don't believe mojo supports Context Managers yet, you will need to call the `close()` method on the file object.

**Reply by rd4com** [reply]
you can use this if you want by syscall  [read into string](https://github.com/Lynet101/Mojo_community-lib/blob/main/IO/c-utilities.mojo) or [ish/shim](https://github.com/lsh/shims/blob/main/read/__init__.mojo)

**Reply by rarebreed** [reply]
Ah nice, I was wondering how to make system calls like an fopen or fstat.  Nice to know about the [external_call intrinsic](https://docs.modular.com/mojo/stdlib/sys/intrinsics.html#external_call)

**Reply by rd4com** [reply]
yes, i got happy when i discovered it too, and it is quite intuitive

**Comment by felixfeixu** [comment]
as for v0.6.1, you can:

```mojo
import pathlib

fn main():
    let p: Path = Path("./test.txt")
    try:
        let str : String = p.read_text()
        print(str)
    finally:
        pass   
```

**Reply by gryznar** [reply]
Is this try...finally needed? It looks like, it does nothing

**Reply by felixfeixu** [reply]
> Is this try...finally needed? It looks like, it does nothing

you can try to remove them and see the compiler errors 😁

**Comment by ghost** [comment]
you can open a file with `with open` now:

```mojo
fn main() raises:
    with open("file", "r") as file:
        print(file.read())
```

**Comment by davidxiaozhi** [comment]
Thanks everyone very much for your solution, which can temporarily solve my problem of reading small files. For large files, I expect to read and process them line by line by reading until the \n character is encountered, instead of reading the entire file at once.
